##########################################################################################
# EPoS project
# Partho Sen: 24.12.2020
# The RNAseq and the Genotypes expression data file sent by Oliver JAN 2020
# The data is corrected for batch, sex and age and also normalized for Steatosis
# Differential Expression of the Genes, detected by the t-test (fast but not accurate) or DESeq (slow but preferable)
# Use DEseq only #
# Analogous script : Make_NAFLD_LiverModels_ver #
# The final DESeq Table can be found in RNAseq_NAFLD folder #
###########################################################################################
library('nlme') ## gives lme & nlme
library('lme4') ## gives lmer
library('gplots')
library('ggplot2')
library('ade4')
library('corrplot')
library('reshape')
library('Hmisc') ## Also using for imputed fumction
library('plyr')
library('car')
library('maptools')
library('rgeos')
library('plotrix')
library('ade4')
library('factoextra')
library('magrittr')
library('DESeq2')
#library('DESeq')
library('readxl')
library('writexl')

rm(list=ls())
setwd('/Users/pasase/OneDrive/Projects_BTK_SysMed_May2017/Partho_NFALD/Results/')

### Read original data and reshape and structure it for the further analysis: Keep off always ###
if(0){
  print('Note this files are generated by Matlab Script: Make_NAFLD_LiverModels_ver.m')
  print('Reading Gene expression files :-')
  
  NAFL_Stea  = read.table('/Users/pasase/OneDrive/Projects_BTK_SysMed_May2017/Partho_NFALD/Data/NAFLD_conditions/NFALD_Expression_NAFL_Stea_genes_formatted.csv',sep=',', header = TRUE,check.names=FALSE) 
  NASH_F01  = read.table('/Users/pasase/OneDrive/Projects_BTK_SysMed_May2017/Partho_NFALD/Data/NAFLD_conditions/NFALD_Expression_NASH_F01_genes_formatted.csv',sep=',', header = TRUE,check.names=FALSE) 
  NASH_F2  = read.table('/Users/pasase/OneDrive/Projects_BTK_SysMed_May2017/Partho_NFALD/Data/NAFLD_conditions/NFALD_Expression_NASH_F2_genes_formatted.csv',sep=',', header = TRUE,check.names=FALSE) 
  NASH_F3  = read.table('/Users/pasase/OneDrive/Projects_BTK_SysMed_May2017/Partho_NFALD/Data/NAFLD_conditions/NFALD_Expression_NASH_F3_genes_formatted.csv',sep=',', header = TRUE,check.names=FALSE) 
  NASH_F4  = read.table('/Users/pasase/OneDrive/Projects_BTK_SysMed_May2017/Partho_NFALD/Data/NAFLD_conditions/NFALD_Expression_NASH_F4_genes_formatted.csv',sep=',', header = TRUE,check.names=FALSE) 
  
  Stea_NASH_F01  = read.table('/Users/pasase/OneDrive/Projects_BTK_SysMed_May2017/Partho_NFALD/Data/NAFLD_conditions/NFALD_Expression_Stea_NASH_F01_genes_formatted.csv',sep=',', header = TRUE,check.names=FALSE) 
  Stea_NASH_F012  = read.table('/Users/pasase/OneDrive/Projects_BTK_SysMed_May2017/Partho_NFALD/Data/NAFLD_conditions/NFALD_Expression_Stea_NASH_F012_genes_formatted.csv',sep=',', header = TRUE,check.names=FALSE) 
  NASH_F23  = read.table('/Users/pasase/OneDrive/Projects_BTK_SysMed_May2017/Partho_NFALD/Data/NAFLD_conditions/NFALD_Expression_NASH_F23_genes_formatted.csv',sep=',', header = TRUE,check.names=FALSE) 
  NASH_F34  = read.table('/Users/pasase/OneDrive/Projects_BTK_SysMed_May2017/Partho_NFALD/Data/NAFLD_conditions/NFALD_Expression_NASH_F34_genes_formatted.csv',sep=',', header = TRUE,check.names=FALSE) 
  
  print('Covert to a suitable dataset/data matrix/dataframe')
  print('Reassigning colnames to sample names, will invert later :-')
  
  rownames(NAFL_Stea) <- NAFL_Stea[,1]; NAFL_Stea <- NAFL_Stea[,-1];
  rownames(NASH_F01)  <- NASH_F01[,1]; NASH_F01 <- NASH_F01[,-1];
  rownames(NASH_F2)   <- NASH_F2[,1]; NASH_F2 <- NASH_F2[,-1];
  rownames(NASH_F3)   <- NASH_F3[,1]; NASH_F3 <- NASH_F3[,-1];
  rownames(NASH_F4)   <- NASH_F4[,1]; NASH_F4 <- NASH_F4[,-1];
  
  rownames(Stea_NASH_F01) <- Stea_NASH_F01[,1]; Stea_NASH_F01 <- Stea_NASH_F01[,-1];
  rownames(Stea_NASH_F012) <- Stea_NASH_F012[,1]; Stea_NASH_F012 <- Stea_NASH_F012[,-1];
  rownames(NASH_F23) <- NASH_F23[,1]; NASH_F23 <- NASH_F23[,-1];
  rownames(NASH_F34) <- NASH_F34[,1]; NASH_F34 <- NASH_F34[,-1];
  
  
  NAFL_Stea = data.matrix(NAFL_Stea);
  NASH_F01 = data.matrix(NASH_F01);
  NASH_F2 = data.matrix(NASH_F2);
  NASH_F3 = data.matrix(NASH_F3);
  NASH_F4 = data.matrix(NASH_F4);
  
  Stea_NASH_F01 = data.matrix(Stea_NASH_F01);
  Stea_NASH_F012 = data.matrix(Stea_NASH_F012);
  NASH_F23 = data.matrix(NASH_F23);
  NASH_F34 = data.matrix(NASH_F34);
  
  colnames(NAFL_Stea) <- c(rep('NAFL_Stea',1,dim(NAFL_Stea)[2])); 
  colnames(NASH_F01) <- c(rep('NASH_F01',1,dim(NASH_F01)[2])); 
  colnames(NASH_F2) <- c(rep('NASH_F2',1,dim(NASH_F2)[2])); 
  colnames(NASH_F3) <- c(rep('NASH_F3',1,dim(NASH_F3)[2])); 
  colnames(NASH_F4) <- c(rep('NASH_F4',1,dim(NASH_F4)[2])); 
  
  colnames(Stea_NASH_F01) <- c( rep('Steatosis_NASH_F01',1,dim(Stea_NASH_F01)[2])); 
  colnames(Stea_NASH_F012) <- c( rep('Steatosis_NASH_F012',1,dim(Stea_NASH_F012)[2])); 
  colnames(NASH_F23) <- c(rep('NASH_F23',1,dim(NASH_F23)[2])); 
  colnames(NASH_F34) <- c(rep('NASH_F34',1,dim(NASH_F34)[2])); 
  
  print('Saving Gene expression RObjects :-')
  
  save(NAFL_Stea , file = "/Users/pasase/OneDrive/Projects_BTK_SysMed_May2017/Partho_NFALD/Data/NAFLD_conditions/NAFL_Stea.rda")
  save(NASH_F01 , file = "/Users/pasase/OneDrive/Projects_BTK_SysMed_May2017/Partho_NFALD/Data/NAFLD_conditions/NASH_F01.rda")
  save(NASH_F2 , file = "/Users/pasase/OneDrive/Projects_BTK_SysMed_May2017/Partho_NFALD/Data/NAFLD_conditions/NASH_F2.rda")
  save(NASH_F3 , file = "/Users/pasase/OneDrive/Projects_BTK_SysMed_May2017/Partho_NFALD/Data/NAFLD_conditions/NASH_F3.rda")
  save(NASH_F4 , file = "/Users/pasase/OneDrive/Projects_BTK_SysMed_May2017/Partho_NFALD/Data/NAFLD_conditions/NASH_F4.rda")
  
  save(Stea_NASH_F01 , file = "/Users/pasase/OneDrive/Projects_BTK_SysMed_May2017/Partho_NFALD/Data/NAFLD_conditions/Stea_NASH_F01.rda")
  save(Stea_NASH_F012 , file = "/Users/pasase/OneDrive/Projects_BTK_SysMed_May2017/Partho_NFALD/Data/NAFLD_conditions/Stea_NASH_F012.rda")
  save(NASH_F23 , file = "/Users/pasase/OneDrive/Projects_BTK_SysMed_May2017/Partho_NFALD/Data/NAFLD_conditions/NASH_F23.rda")
  save(NASH_F34 , file = "/Users/pasase/OneDrive/Projects_BTK_SysMed_May2017/Partho_NFALD/Data/NAFLD_conditions/NASH_F34.rda")
  
  print('done!')
  
}

### Shape the data frame and save and load for the further analysis: Keep off ###
if(1){
  print('loading files ...')
  load(file = "/Users/pasase/OneDrive/Projects_BTK_SysMed_May2017/Partho_NFALD/Data/NAFLD_conditions/NAFL_Stea.rda")
  load(file = "/Users/pasase/OneDrive/Projects_BTK_SysMed_May2017/Partho_NFALD/Data/NAFLD_conditions/NASH_F01.rda")
  load(file = "/Users/pasase/OneDrive/Projects_BTK_SysMed_May2017/Partho_NFALD/Data/NAFLD_conditions/NASH_F2.rda")
  load(file = "/Users/pasase/OneDrive/Projects_BTK_SysMed_May2017/Partho_NFALD/Data/NAFLD_conditions/NASH_F3.rda")
  load(file = "/Users/pasase/OneDrive/Projects_BTK_SysMed_May2017/Partho_NFALD/Data/NAFLD_conditions/NASH_F4.rda")
  
  load(file = "/Users/pasase/OneDrive/Projects_BTK_SysMed_May2017/Partho_NFALD/Data/NAFLD_conditions/Stea_NASH_F01.rda")
  load(file = "/Users/pasase/OneDrive/Projects_BTK_SysMed_May2017/Partho_NFALD/Data/NAFLD_conditions/Stea_NASH_F012.rda")
  load(file = "/Users/pasase/OneDrive/Projects_BTK_SysMed_May2017/Partho_NFALD/Data/NAFLD_conditions/NASH_F23.rda")
  load(file = "/Users/pasase/OneDrive/Projects_BTK_SysMed_May2017/Partho_NFALD/Data/NAFLD_conditions/NASH_F34.rda")
  print('done!')
  
}


#########################################################################################################################################################
### load the data object [DFF] with genes and groups for the differential expression analysis ###
#########################################################################################################################################################

### ******** PCA *************** ###
if(TRUE){
  source("/Users/pasase/OneDrive/Projects_BTK_SysMed_May2017/Partho_Celiac3_new/Analysis/range.R")
  source("/Users/pasase/OneDrive/Projects_BTK_SysMed_May2017/Partho_Celiac3_new/Analysis/ellipseplot.R")
  source("/Users/pasase/OneDrive/Projects_BTK_SysMed_May2017/Partho_Celiac3_new/Analysis/PCAvarAxis.R")
  
  # DFF1 = cbind(Stea_NASH_F01,Stea_NASH_F012,Stea_NASH_F0123,
  #              NASH_F23,NASH_F34,NASH_F4)
  
  DFF1 = cbind(NAFL_Stea,NASH_F01,NASH_F2,NASH_F3,NASH_F4,Stea_NASH_F01,Stea_NASH_F012,NASH_F23,NASH_F34)
  A2 = t(DFF1);
  PCs <- prcomp(A2, scale = TRUE, center = TRUE); ## eigs <- pca$sdev^2
  
  library('randomcoloR')
  col3 <- distinctColorPalette(9)
  
  explainPCAvar <- PCAvarAxis(PCs)
  
  pdf("./Figure_new/PCA_RNAseq_corrected_NAFLD_Oliver_4Grps.pdf", width = 12, height = 9,paper = 'special')
  oripar=par()$mar
  #par(xpd=F,mar=oripar+c(2,2,0,12)) ## c(down,left,up,right) take some area to right # XPD = T plots outside
  par(xpd=T, mar=oripar+c(2,2,0,6))
  
  ellipseplot(PCs$x[,1],                              # data for x-axis
              PCs$x[,2],                              # data for y-axis
              as.factor(rownames(PCs$x)),                        # factor with classes
              pcol=col3,            # colors for plotting (must match # of factors) ; '#F01F02','#04F1F9''#999999','#FFE4C4)
              pbgcol=FALSE,                           # point borders black?
              cexsize=2,                              # size of points 
              ppch=rep(21,9),                         # shape of points (must match # of factors)
              legpos="topright",                      # position of legend           
              legcexsize=1,                          # legend text size
              legptsize=2,                           # legend point size 
              axissize=1,                            # Set axis text size
              linewidth=2                            # Set axis line size
  )  
  #draw.ellipse(PCs$x[,1],PCs$x[,2])
  #grid(lwd=1)
  
  #text(PCs$x[,1],                              # data for x-axis
  #PCs$x[,2],
  #labels = rownames(A),
  #cex = 0.6)
  
  title(xlab=explainPCAvar[["PC1"]],    # % variance explained on PC1
        ylab=explainPCAvar[["PC2"]],    # % variance explained on PC2 
        main="PCA",                  # Title
        cex.lab=2,                    # size of label text
        cex.main=2                    # size of title text
  )
  #summary(PCs)# it has variance info
  dev.off()
  
}

########################################## **** Differential expression analysis **** ##########################################
######################################## ********* *DESeq2 analysis* ********* #################

#DFF1 = cbind(Stea_NASH_F01,NASH_F23,NASH_F4,HCs)
#DFF1 = cbind(Stea_NASH_F01,Stea_NASH_F012,NASH_F23,NASH_F34,NASH_F24);
DFF1 = cbind(NAFL_Stea,NASH_F01,NASH_F2,NASH_F3,NASH_F4,Stea_NASH_F01,Stea_NASH_F012,NASH_F23,NASH_F34)

if(TRUE){
  
  if(0){
    print('DESeq2 selected::Estimation takes a while');
    print('Make DEFSeq2 object:-');
    
    dds <- DESeqDataSetFromMatrix(round(as.data.frame(2^DFF1),0),data.frame(Group=as.factor(colnames(DFF1))), ~ Group)
    #dds <- estimateSizeFactors(dds)
    print(dim(dds));
    
    
    if(0){ ## Instead go for full gene expression # time expensive ## less reporter metabolites found
          print('Filter the genes for making it faster ... not the sample')
          nsamp <- 422 ## 70% sample
          keep <- rowSums(counts(dds, normalized=T) >= 10) >= nsamp;
          table(keep);
          dds <- dds[keep,]; dim(dds);
           #head(counts(dds, normalized=T))
          hist(head(counts(dds,normalized=T)),10);
    }
    

    
    print('Perform analysis ...')
        dds.NFALD.Condi <- DESeq(dds);
    print('done!');
    
    print('Saving diffExp object ....')
    save(dds.NFALD.Condi, file = "/Users/pasase/OneDrive/Projects_BTK_SysMed_May2017/Partho_NFALD/Data/RNAseq_NAFLD/dds.NFALD.Condi.rda");
    print('done!')
  }
  
  
  ################################################################################################################
  ##################################### FETCH the differentially expressed genes for each constrast ##############
  ################################################################################################################
  
  if(TRUE){
    
    load(file = "/Users/pasase/OneDrive/Projects_BTK_SysMed_May2017/Partho_NFALD/Data/RNAseq_NAFLD/dds.NFALD.Condi.rda"); ## Object is dds #
    
    #### ** Decided on the metting ** ####
    
    ## Category:G1
    print('Performing Differential Expression Analysis by DESeq2: G1');
    DF.NASH_F01.NAFL_Stea <- results(dds.NFALD.Condi,cooksCutoff = 0.99,contrast=c('Group','NASH_F01','NAFL_Stea'),independentFiltering = TRUE, alpha = 0.05, pAdjustMethod = "BH"); ## Supply many grps or constrast as lists ##
    ## Category:G2
    print('Performing Differential Expression Analysis by DESeq2: G2');
    DF.NASH_F2.NAFL_Stea <- results(dds.NFALD.Condi,cooksCutoff = 0.99,contrast=c('Group','NASH_F2','NAFL_Stea'),independentFiltering = TRUE, alpha = 0.05, pAdjustMethod = "BH"); ## Supply many grps or constrast as lists ##
    ## Category:G3
    print('Performing Differential Expression Analysis by DESeq2: G3');
    DF.NASH_F3.NAFL_Stea <- results(dds.NFALD.Condi,cooksCutoff = 0.99,contrast=c('Group','NASH_F3','NAFL_Stea'),independentFiltering = TRUE, alpha = 0.05, pAdjustMethod = "BH"); ## Supply many grps or constrast as lists ##
    ## Category:G4
    print('Performing Differential Expression Analysis by DESeq2: G4');
    DF.NASH_F4.NAFL_Stea <- results(dds.NFALD.Condi,cooksCutoff = 0.99,contrast=c('Group','NASH_F4','NAFL_Stea'),independentFiltering = TRUE, alpha = 0.05, pAdjustMethod = "BH"); ## Supply many grps or constrast as lists ##
    
    
    ## Category:G5
    print('Performing Differential Expression Analysis by DESeq2: G5');
    DF.F23.Stea.F01 <- results(dds.NFALD.Condi,cooksCutoff = 0.99,contrast=c('Group','NASH_F23','Steatosis_NASH_F01'),independentFiltering = TRUE, alpha = 0.05, pAdjustMethod = "BH"); ## Supply many grps or constrast as lists ##
    ## Category:G6
    print('Performing Differential Expression Analysis by DESeq2: G6');
    DF.F34.Stea.F01 <- results(dds.NFALD.Condi,cooksCutoff = 0.99,contrast=c('Group','NASH_F34','Steatosis_NASH_F01'),independentFiltering = TRUE, alpha = 0.05, pAdjustMethod = "BH"); ## Supply many grps or constrast as lists ##
    ## Category:G7
    print('Performing Differential Expression Analysis by DESeq2: G7');
    DF.F34.Stea.F012 <- results(dds.NFALD.Condi,cooksCutoff = 0.99,contrast=c('Group','NASH_F34','Steatosis_NASH_F012'),independentFiltering = TRUE, alpha = 0.05, pAdjustMethod = "BH"); ## Supply many grps or constrast as lists ##
    
  }
  
} ## Turn off the Loop for DESeq2 ##


############ Put differential expression in object for the volcano plots ###################################################################
print('Preparing for volcano plots');
#source('/Users/pasase/OneDrive/Projects_BTK_SysMed_May2017/Partho_NFALD/Scripts/volcano2.R')

de_threshold=0.05; ## FDR corrected p-values 
res=list(); res1=list(); nom=list();

nom[[1]]  = "NASH_F01 vs. NAFL_Stea";
res[[1]]  = DF.NASH_F01.NAFL_Stea; ## Original data
res1[[1]] = subset(DF.NASH_F01.NAFL_Stea, padj < de_threshold & abs(log2FoldChange) > 0);    ## subsetting data

nom[[2]]  = "NASH_F2 vs. NAFL_Stea";
res[[2]]  = DF.NASH_F2.NAFL_Stea; ## Original data
res1[[2]] = subset(DF.NASH_F2.NAFL_Stea, padj < de_threshold & abs(log2FoldChange) > 0);    ## subsetting data

nom[[3]]  = "NASH_F3 vs. NAFL_Stea";
res[[3]]  = DF.NASH_F3.NAFL_Stea; ## Original data
res1[[3]] = subset(DF.NASH_F3.NAFL_Stea, padj < de_threshold & abs(log2FoldChange) > 0);    ## subsetting data

nom[[4]]  = "NASH_F4 vs. NAFL_Stea";
res[[4]]  = DF.NASH_F4.NAFL_Stea; ## Original data
res1[[4]] = subset(DF.NASH_F4.NAFL_Stea, padj < de_threshold & abs(log2FoldChange) > 0);    ## subsetting data


nom[[5]]  = "NASH_F23 vs. Steatosis_NASH_F01";
res[[5]]  = DF.F23.Stea.F01; ## Original data
res1[[5]] = subset(DF.F23.Stea.F01, padj < de_threshold & abs(log2FoldChange) > 0);    ## subsetting data

nom[[6]]  = 'NASH_F34 vs. Steatosis_NASH_F01';
res[[6]]  = DF.F34.Stea.F01; ## Original data
res1[[6]] = subset(DF.F34.Stea.F01, padj < de_threshold & abs(log2FoldChange) > 0);    ## subsetting data

nom[[7]]  = 'NASH_F34 vs. Steatosis_NASH_F012';
res[[7]]  = DF.F34.Stea.F012; ## Original data
res1[[7]] = subset(DF.F34.Stea.F012, padj < de_threshold & abs(log2FoldChange) > 0);    ## subsetting data

############ Volcano plot card in the loop ###################################################################
print('Make volcano plots.....');

pdf("./Figure_new/VolcanoPlot_RNAseq_corrected_NAFLD_Oliver.pdf", width = 15, height = 10)
oripar=par()$mar
par(xpd=F, mar=oripar+c(2,2,0,6)) ## c(down,left,up,right) take some area to right # XPD = T plots outside
par(mfrow=c(3,3))

for (ii in 1:length(res)) {
  
  obj = res[[ii]];
  obj1 = res1[[ii]];
  
  with(obj, plot(log2FoldChange, -log10(padj), pch=21, main=nom[[ii]], col="gray60",bg='gray90',xlim=c(-3,3), ylim=c(0,20) ))
  with(subset(obj1, padj < de_threshold & log2FoldChange > 0), points(log2FoldChange, -log10(padj), pch=21, col="red",bg='orange'))
  with(subset(obj1, padj < de_threshold & log2FoldChange < 0), points(log2FoldChange, -log10(padj), pch=21, col="blue",bg='darkturquoise'))
  
  abline(h = -log10(de_threshold), col="black", lwd=1, lty=2)
  abline(v = 0, col="black", lwd=1, lty=2)
  
  a=dim(subset(obj1, padj < de_threshold & log2FoldChange > 0))[1] ;
  b=dim(subset(obj1, padj < de_threshold & log2FoldChange < 0))[1] ;
  
  legend('topright',legend=c(paste('up:',a), paste('down:',b)), pch = c(21,21), pt.bg = c("orange","darkturquoise"), pt.cex=1.5,border = NA, cex=0.8);
  
}

dev.off()

################ *Compare DEG among three different conditions* #################################

flgs.comparision = c("NASH(F0-1) vs. NAFL","NASH(F2) vs. NAFL","NASH(F3) vs. NAFL","NASH(F4) vs. NAFL",
                     "NASH(F2-3) vs. NASH(F0-1)", "NASH(F3-4) vs. NAFL+NASH(F0-1)", "NASH(F3-4) vs. NAFL+NASH(F0-2)");

### ******* write to a file ******** ####################
print('Write differential expression results to a file.....');

Diff.Exp.NFALD.DESeq2 <- lapply(res1,as.data.frame);
save(Diff.Exp.NFALD.DESeq2 , file = "/Users/pasase/OneDrive/Projects_BTK_SysMed_May2017/Partho_NFALD/Data/RNAseq_NAFLD/Diff.Exp.NFALD.DESeq2.rda");
load(file = "/Users/pasase/OneDrive/Projects_BTK_SysMed_May2017/Partho_NFALD/Data/RNAseq_NAFLD/Diff.Exp.NFALD.DESeq2.rda")

print('Appending rownames of the list as a column.....')

Mat=list();
for(uu in 1:length(Diff.Exp.NFALD.DESeq2)){
  
  Mat[[uu]] = cbind('Genes' = rownames(Diff.Exp.NFALD.DESeq2[[uu]]), Diff.Exp.NFALD.DESeq2[[uu]]) 
  
}


names(Mat) <- flgs.comparision;
write_xlsx(Mat,"/Users/pasase/OneDrive/Projects_BTK_SysMed_May2017/Partho_NFALD/Data/RNAseq_NAFLD/NAFLD_differential_expression_DSeq2_ver2.0.xlsx"); ## works with t-test switch but preferably DEseq2
print('done!')

#################################################################### *************************** The END *********************** #############################################################################